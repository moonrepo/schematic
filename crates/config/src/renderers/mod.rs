/// Renders TypeScript types.
#[cfg(feature = "typescript")]
pub mod typescript;
