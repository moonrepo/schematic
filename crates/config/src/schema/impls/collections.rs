// use crate::config::ConfigSchema;
// use crate::schema::types::*;

// impl<T: ConfigSchema> ConfigSchema for Vec<T> {
//     fn generate_schema() -> Schema {
//         let inner = T::generate_schema();

//         if let Schema::Type {
//             ref mut nullable, ..
//         } = schema
//         {
//             *nullable = true;
//         }

//         schema
//     }
// }
