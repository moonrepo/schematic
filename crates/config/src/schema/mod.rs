mod generator;
mod renderer;

pub use generator::*;
pub use indexmap::*;
pub use renderer::*;
pub use schematic_types::*;
