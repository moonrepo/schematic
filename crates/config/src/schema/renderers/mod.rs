#[cfg(feature = "json_schema")]
pub mod json_schema;

#[cfg(feature = "template")]
pub mod template;

#[cfg(feature = "typescript")]
pub mod typescript;
