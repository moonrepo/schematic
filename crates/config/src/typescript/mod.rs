mod converter;
mod generator;

pub use converter::*;
pub use generator::*;
