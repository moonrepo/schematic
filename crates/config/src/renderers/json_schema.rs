use crate::schema::{RenderResult, SchemaRenderer};
use schemars::schema::*;
use schematic_types::*;
use std::collections::{BTreeMap, BTreeSet, HashSet};

#[derive(Default)]
pub struct JsonSchemaOptions {}

/// Renders JsonSchema types from a schema.
#[derive(Default)]
pub struct JsonSchemaRenderer {
    options: JsonSchemaOptions,
    references: HashSet<String>,
}

impl JsonSchemaRenderer {
    pub fn new(options: JsonSchemaOptions) -> Self {
        Self {
            options,
            references: HashSet::new(),
        }
    }

    fn create_metadata(&self, name: Option<&String>) -> Option<Box<Metadata>> {
        if let Some(name) = name {
            Some(Box::new(Metadata {
                title: Some(name.to_owned()),
                ..Default::default()
            }))
        } else {
            None
        }
    }

    fn create_schema_from_field(&self, field: &SchemaField) -> RenderResult<Schema> {
        let mut schema = self.render_schema(&field.type_of)?;

        if let Schema::Object(ref mut inner) = schema {
            inner.metadata = Some(Box::new(Metadata {
                title: field.name.clone(),
                description: field.description.clone(),
                deprecated: field.deprecated,
                read_only: field.read_only,
                write_only: field.write_only,
                ..Default::default()
            }));
        }

        Ok(schema)
    }
}

impl SchemaRenderer<Schema> for JsonSchemaRenderer {
    fn is_reference(&self, name: &str) -> bool {
        self.references.contains(name)
    }

    fn render_array(&self, array: &ArrayType) -> RenderResult<Schema> {
        let data = SchemaObject {
            instance_type: Some(SingleOrVec::Single(Box::new(InstanceType::Array))),
            metadata: self.create_metadata(array.name.as_ref()),
            array: Some(Box::new(ArrayValidation {
                items: Some(SingleOrVec::Single(Box::new(
                    self.render_schema(&array.items_type)?,
                ))),
                max_items: array.max_length.map(|i| i as u32),
                min_items: array.min_length.map(|i| i as u32),
                unique_items: Some(array.unique),
                ..Default::default()
            })),
            ..Default::default()
        };

        Ok(Schema::Object(data))
    }

    fn render_boolean(&self) -> RenderResult<Schema> {
        Ok(Schema::Object(SchemaObject {
            instance_type: Some(SingleOrVec::Single(Box::new(InstanceType::Boolean))),
            ..Default::default()
        }))
    }

    fn render_float(&self, float: &FloatType) -> RenderResult<Schema> {
        let data = SchemaObject {
            instance_type: Some(SingleOrVec::Single(Box::new(InstanceType::Number))),
            metadata: self.create_metadata(float.name.as_ref()),
            format: float.format.clone(),
            number: Some(Box::new(NumberValidation {
                exclusive_maximum: float.max_exclusive,
                exclusive_minimum: float.min_exclusive,
                maximum: float.max,
                minimum: float.min,
                multiple_of: float.multiple_of,
            })),
            ..Default::default()
        };

        Ok(Schema::Object(data))
    }

    fn render_integer(&self, integer: &IntegerType) -> RenderResult<Schema> {
        let data = SchemaObject {
            instance_type: Some(SingleOrVec::Single(Box::new(InstanceType::Number))),
            metadata: self.create_metadata(integer.name.as_ref()),
            format: integer.format.clone(),
            number: Some(Box::new(NumberValidation {
                exclusive_maximum: integer.max_exclusive.map(|i| i as f64),
                exclusive_minimum: integer.min_exclusive.map(|i| i as f64),
                maximum: integer.max.map(|i| i as f64),
                minimum: integer.min.map(|i| i as f64),
                multiple_of: integer.multiple_of.map(|i| i as f64),
            })),
            ..Default::default()
        };

        Ok(Schema::Object(data))
    }

    fn render_literal(&self, literal: &LiteralType) -> RenderResult<Schema> {
        // TODO?
        // if let Some(value) = &literal.value {
        //     return Ok(match value {
        //         LiteralValue::Bool(inner) => inner.to_string(),
        //         LiteralValue::Float(inner) => inner.to_owned(),
        //         LiteralValue::Int(inner) => inner.to_string(),
        //         LiteralValue::UInt(inner) => inner.to_string(),
        //         LiteralValue::String(inner) => format!("'{inner}'"),
        //     });
        // }

        self.render_unknown()
    }

    fn render_null(&self) -> RenderResult<Schema> {
        Ok(Schema::Object(SchemaObject {
            instance_type: Some(SingleOrVec::Single(Box::new(InstanceType::Null))),
            ..Default::default()
        }))
    }

    fn render_object(&self, object: &ObjectType) -> RenderResult<Schema> {
        let data = SchemaObject {
            instance_type: Some(SingleOrVec::Single(Box::new(InstanceType::Object))),
            metadata: self.create_metadata(object.name.as_ref()),
            object: Some(Box::new(ObjectValidation {
                max_properties: object.max_fields.map(|i| i as u32),
                min_properties: object.min_fields.map(|i| i as u32),
                required: BTreeSet::from_iter(object.required.clone()),
                additional_properties: Some(Box::new(self.render_schema(&object.value_type)?)),
                property_names: Some(Box::new(self.render_schema(&object.key_type)?)),
                ..Default::default()
            })),
            ..Default::default()
        };

        Ok(Schema::Object(data))
    }

    fn render_reference(&self, reference: &str) -> RenderResult<Schema> {
        Ok(Schema::Object(SchemaObject {
            reference: Some(reference.into()),
            ..Default::default()
        }))
    }

    fn render_string(&self, string: &StringType) -> RenderResult<Schema> {
        let data = SchemaObject {
            instance_type: Some(SingleOrVec::Single(Box::new(InstanceType::String))),
            metadata: self.create_metadata(string.name.as_ref()),
            format: string.format.clone(),
            string: Some(Box::new(StringValidation {
                max_length: string.max_length.map(|i| i as u32),
                min_length: string.min_length.map(|i| i as u32),
                pattern: string.pattern.clone(),
            })),
            ..Default::default()
        };

        Ok(Schema::Object(data))
    }

    fn render_struct(&self, structure: &StructType) -> RenderResult<Schema> {
        let mut properties = BTreeMap::new();
        let mut required = BTreeSet::from_iter(structure.required.clone());

        for field in &structure.fields {
            let name = field.name.clone().unwrap();

            if !field.optional {
                required.insert(name.clone());
            }

            properties.insert(name, self.create_schema_from_field(field)?);
        }

        let data = SchemaObject {
            instance_type: Some(SingleOrVec::Single(Box::new(InstanceType::Object))),
            metadata: self.create_metadata(structure.name.as_ref()),
            object: Some(Box::new(ObjectValidation {
                required,
                properties,
                ..Default::default()
            })),
            ..Default::default()
        };

        Ok(Schema::Object(data))
    }

    fn render_tuple(&self, tuple: &TupleType) -> RenderResult<Schema> {
        let mut items = vec![];

        for item in &tuple.items_types {
            items.push(self.render_schema(item)?);
        }

        let data = SchemaObject {
            instance_type: Some(SingleOrVec::Single(Box::new(InstanceType::Array))),
            metadata: self.create_metadata(tuple.name.as_ref()),
            array: Some(Box::new(ArrayValidation {
                items: Some(SingleOrVec::Vec(items)),
                max_items: Some(tuple.items_types.len() as u32),
                min_items: Some(tuple.items_types.len() as u32),
                ..Default::default()
            })),
            ..Default::default()
        };

        Ok(Schema::Object(data))
    }

    fn render_union(&self, uni: &UnionType) -> RenderResult<Schema> {
        // let mut items = vec![];

        // for item in &uni.variants_types {
        //     items.push(self.render_schema(item)?);
        // }

        // Ok(items.join(" | "))

        self.render_unknown()
    }

    fn render_unknown(&self) -> RenderResult<Schema> {
        Ok(Schema::Object(SchemaObject {
            instance_type: Some(SingleOrVec::Single(Box::new(InstanceType::Null))),
            ..Default::default()
        }))
    }

    fn render(&mut self, schemas: &[SchemaType], references: &HashSet<String>) -> RenderResult {
        self.references.extend(references.to_owned());

        let mut outputs = vec![
            "// Automatically generated by schematic. DO NOT MODIFY!".to_string(),
            "/* eslint-disable */".to_string(),
        ];

        for schema in schemas {
            // if let Some(name) = schema.get_name() {
            //     outputs.push(match schema {
            //         SchemaType::Struct(inner) => self.export_object_type(name, inner)?,
            //         SchemaType::Union(inner) => self.export_enum_type(name, inner)?,
            //         _ => self.export_type_alias(name, self.render_schema(schema)?)?,
            //     });
            // }
        }

        Ok(outputs.join("\n\n"))
    }
}
