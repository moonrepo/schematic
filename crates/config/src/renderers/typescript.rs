use crate::schema::{RenderResult, SchemaRenderer};
use schematic_types::*;
use std::collections::HashSet;

/// Format of a TypeScript enum.
#[derive(Default)]
pub enum EnumFormat {
    /// Native enum: `enum Name { Foo, Bar }`
    Enum,
    /// Native enum with values: `enum Name { Foo = 'foo', Bar = 'bar' }`
    ValuedEnum,
    /// String union: `type Name = 'foo' | 'bar';`
    #[default]
    Union,
}

/// Format of a TypeScript object.
#[derive(Default)]
pub enum ObjectFormat {
    /// Interface: `interface Foo {}`
    #[default]
    Interface,
    /// Type alias: `type Foo = {};`
    Type,
}

/// Options to control the rendered TypeScript output.
#[derive(Default)]
pub struct TypeScriptOptions {
    pub const_enum: bool,
    pub enum_format: EnumFormat,
    pub object_format: ObjectFormat,
}

/// Renders TypeScript types from a schema.
#[derive(Default)]
pub struct TypeScriptRenderer {
    options: TypeScriptOptions,
    references: HashSet<String>,
}

impl TypeScriptRenderer {
    pub fn new(options: TypeScriptOptions) -> Self {
        Self {
            options,
            references: HashSet::new(),
        }
    }

    fn export_type_alias(&self, name: &str, value: String) -> RenderResult {
        Ok(format!("export type {} = {};", name, value))
    }

    fn export_enum_type(&self, name: &str, enu: &EnumType) -> RenderResult {
        let value = self.render_enum(enu)?;
        let out = format!("enum {} {{\n{}\n}}", name, value);

        Ok(if self.options.const_enum {
            format!("export const {}", out)
        } else {
            format!("export {}", out)
        })
    }

    fn export_object_type(&self, name: &str, structure: &StructType) -> RenderResult {
        let value = self.render_struct(structure)?;

        if matches!(self.options.object_format, ObjectFormat::Interface) {
            return Ok(format!("export interface {} {}", name, value));
        }

        self.export_type_alias(name, value)
    }

    fn wrap_in_comment(&self, comment: &str, value: String) -> String {
        if comment.starts_with('*') {
            let mut out = vec!["\t/**".to_owned()];

            for line in comment.split('\n') {
                out.push(format!("\t {}", line.trim()));
            }

            out.push("\t */".to_owned());

            format!("{}\n{}", out.join("\n"), value)
        } else {
            format!("\t// {}\n{}", comment.trim(), value)
        }
    }
}

impl SchemaRenderer<String> for TypeScriptRenderer {
    fn is_reference(&self, name: &str) -> bool {
        self.references.contains(name)
    }

    fn render_array(&self, array: &ArrayType) -> RenderResult {
        Ok(format!("{}[]", self.render_schema(&array.items_type)?))
    }

    fn render_boolean(&self) -> RenderResult {
        Ok("boolean".into())
    }

    fn render_enum(&self, enu: &EnumType) -> RenderResult {
        let mut fields = vec![];

        for variant in enu.variants.as_ref().unwrap() {
            if variant.hidden {
                continue;
            }

            if let Some(variant_name) = &variant.name {
                let mut field = if matches!(self.options.enum_format, EnumFormat::ValuedEnum) {
                    format!(
                        "\t{} = {},",
                        variant_name,
                        self.render_schema(&variant.type_of)?
                    )
                } else {
                    format!("\t{},", variant_name)
                };

                if let Some(comment) = &variant.description {
                    field = self.wrap_in_comment(comment.trim(), field);
                }

                fields.push(field);
            }
        }

        Ok(fields.join("\n"))
    }

    fn render_float(&self, _: &FloatType) -> RenderResult {
        Ok("number".into())
    }

    fn render_integer(&self, _: &IntegerType) -> RenderResult {
        Ok("number".into())
    }

    fn render_literal(&self, literal: &LiteralType) -> RenderResult {
        if let Some(value) = &literal.value {
            return Ok(match value {
                LiteralValue::Bool(inner) => inner.to_string(),
                LiteralValue::Float(inner) => inner.to_owned(),
                LiteralValue::Int(inner) => inner.to_string(),
                LiteralValue::UInt(inner) => inner.to_string(),
                LiteralValue::String(inner) => format!("'{inner}'"),
            });
        }

        self.render_unknown()
    }

    fn render_null(&self) -> RenderResult {
        Ok("null".into())
    }

    fn render_object(&self, object: &ObjectType) -> RenderResult {
        Ok(format!(
            "Record<{}, {}>",
            self.render_schema(&object.key_type)?,
            self.render_schema(&object.value_type)?
        ))
    }

    fn render_reference(&self, reference: &str) -> RenderResult {
        Ok(reference.into())
    }

    fn render_string(&self, _: &StringType) -> RenderResult {
        Ok("string".into())
    }

    fn render_struct(&self, structure: &StructType) -> RenderResult {
        let mut out = vec![];

        for field in &structure.fields {
            if field.hidden {
                continue;
            }

            let mut row = format!("\t{}", field.name.as_ref().unwrap());

            if field.optional {
                row.push_str("?: ");
            } else {
                row.push_str(": ");
            }

            row.push_str(&self.render_schema(&field.type_of)?);

            if field.nullable && !row.contains(" null") {
                row.push_str(" | null");
            }

            if matches!(self.options.object_format, ObjectFormat::Interface) {
                row.push(';');
            } else {
                row.push(',');
            }

            if let Some(comment) = &field.description {
                row = self.wrap_in_comment(comment.trim(), row);
            }

            out.push(row);
        }

        Ok(format!("{{\n{}\n}}", out.join("\n")))
    }

    fn render_tuple(&self, tuple: &TupleType) -> RenderResult {
        let mut items = vec![];

        for item in &tuple.items_types {
            items.push(self.render_schema(item)?);
        }

        Ok(format!("[{}]", items.join(", ")))
    }

    fn render_union(&self, uni: &UnionType) -> RenderResult {
        let mut items = vec![];

        for item in &uni.variants_types {
            items.push(self.render_schema(item)?);
        }

        Ok(items.join(" | "))
    }

    fn render_unknown(&self) -> RenderResult {
        Ok("unknown".into())
    }

    fn render(&mut self, schemas: &[SchemaType], references: &HashSet<String>) -> RenderResult {
        self.references.extend(references.to_owned());

        let mut outputs = vec![
            "// Automatically generated by schematic. DO NOT MODIFY!".to_string(),
            "/* eslint-disable */".to_string(),
        ];

        for schema in schemas {
            if let Some(name) = schema.get_name() {
                outputs.push(match schema {
                    SchemaType::Enum(inner) => {
                        if matches!(self.options.enum_format, EnumFormat::Union)
                            || inner.variants.is_none()
                        {
                            self.export_type_alias(
                                name,
                                self.render_union(&UnionType {
                                    name: inner.name.clone(),
                                    variants_types: inner
                                        .values
                                        .iter()
                                        .map(|v| Box::new(SchemaType::Literal(v.clone())))
                                        .collect(),
                                    variants: inner.variants.clone(),
                                    ..Default::default()
                                })?,
                            )?
                        } else {
                            self.export_enum_type(name, inner)?
                        }
                    }
                    SchemaType::Struct(inner) => self.export_object_type(name, inner)?,
                    _ => self.export_type_alias(name, self.render_schema(schema)?)?,
                });
            }
        }

        Ok(outputs.join("\n\n"))
    }
}
