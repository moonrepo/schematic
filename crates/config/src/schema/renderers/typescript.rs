use crate::schema::{RenderResult, SchemaRenderer};
use indexmap::IndexMap;
use schematic_types::*;
use std::collections::{HashMap, HashSet};

/// Format of a TypeScript enum.
#[derive(Default)]
pub enum EnumFormat {
    /// Native enum: `enum Name { Foo, Bar }`
    Enum,
    /// Native enum with values: `enum Name { Foo = 'foo', Bar = 'bar' }`
    ValuedEnum,
    /// String union: `type Name = 'foo' | 'bar';`
    #[default]
    Union,
}

/// Format of a TypeScript object.
#[derive(Default)]
pub enum ObjectFormat {
    /// Interface: `interface Foo {}`
    #[default]
    Interface,
    /// Type alias: `type Foo = {};`
    Type,
}

/// Options to control the rendered TypeScript output.
#[derive(Default)]
pub struct TypeScriptOptions {
    /// Render a `const enum` instead of a `enum`.
    pub const_enum: bool,

    /// Disable references and render all types inline recursively.
    pub disable_references: bool,

    /// Format to render enums, either an `enum` or a `type` union.
    pub enum_format: EnumFormat,

    /// List of references to exclude from exporting as a type.
    pub exclude_references: HashSet<String>,

    /// Map of relative import file paths to a list of types to import.
    /// Will be rendered as an `import type {} from 'path';` declaration.
    pub external_types: HashMap<String, HashSet<String>>,

    /// Character(s) to use for indentation.
    pub indent_char: String,

    /// Format to render objects, either an `interface` or `type`.
    pub object_format: ObjectFormat,
}

/// Renders TypeScript types from a schema.
#[derive(Default)]
pub struct TypeScriptRenderer {
    depth: usize,
    options: TypeScriptOptions,
    references: HashSet<String>,
}

impl TypeScriptRenderer {
    pub fn new(options: TypeScriptOptions) -> Self {
        Self {
            depth: 0,
            options,
            references: HashSet::new(),
        }
    }

    fn indent(&self) -> String {
        let chars = if self.options.indent_char.is_empty() {
            "\t"
        } else {
            &self.options.indent_char
        };

        if self.depth == 0 {
            String::new()
        } else {
            chars.repeat(self.depth)
        }
    }

    fn is_excluded(&self, name: &str) -> bool {
        self.options.exclude_references.contains(name)
    }

    fn is_external(&self, name: &str) -> bool {
        for externals in self.options.external_types.values() {
            if externals.contains(name) {
                return true;
            }
        }

        false
    }

    fn is_string_union_enum(&self, enu: &EnumType) -> bool {
        matches!(self.options.enum_format, EnumFormat::Union)
            || enu.variants.is_none()
            || self.options.disable_references
    }

    fn export_type_alias(&mut self, name: &str, value: String) -> RenderResult {
        Ok(format!("export type {} = {};", name, value))
    }

    fn export_enum_type(&mut self, name: &str, enu: &EnumType) -> RenderResult {
        let value = self.render_enum(enu)?;

        let output = if self.is_string_union_enum(enu) {
            self.export_type_alias(name, value)?
        } else {
            let out = format!("enum {} {}", name, value);

            if self.options.const_enum {
                format!("export const {}", out)
            } else {
                format!("export {}", out)
            }
        };

        Ok(self.wrap_in_comment(enu.description.as_ref(), None, output))
    }

    fn export_object_type(&mut self, name: &str, structure: &StructType) -> RenderResult {
        let value = self.render_struct(structure)?;

        let output = if matches!(self.options.object_format, ObjectFormat::Interface) {
            format!("export interface {} {}", name, value)
        } else {
            self.export_type_alias(name, value)?
        };

        Ok(self.wrap_in_comment(structure.description.as_ref(), None, output))
    }

    fn render_enum_or_union(&mut self, enu: &EnumType) -> RenderResult {
        if self.is_string_union_enum(enu) {
            return self.render_union(&UnionType {
                variants_types: enu
                    .values
                    .iter()
                    .map(|v| Box::new(SchemaType::Literal(v.clone())))
                    .collect(),
                variants: enu.variants.clone(),
                ..Default::default()
            });
        }

        self.depth += 1;

        let mut out = vec![];
        let indent = self.indent();

        for variant in enu.variants.as_ref().unwrap() {
            if variant.hidden {
                continue;
            }

            if let Some(variant_name) = &variant.name {
                let mut field = if matches!(self.options.enum_format, EnumFormat::ValuedEnum) {
                    format!(
                        "{}{} = {},",
                        indent,
                        variant_name,
                        self.render_schema(&variant.type_of)?
                    )
                } else {
                    format!("{}{},", indent, variant_name)
                };

                field = self.wrap_in_comment(
                    variant.description.as_ref(),
                    variant.type_of.get_default(),
                    field,
                );

                out.push(field);
            }
        }

        self.depth -= 1;

        Ok(format!("{{\n{}\n{}}}", out.join("\n"), self.indent()))
    }

    fn lit_to_string(&self, lit: &LiteralValue) -> String {
        match lit {
            LiteralValue::Bool(inner) => inner.to_string(),
            LiteralValue::F32(inner) => inner.to_string(),
            LiteralValue::F64(inner) => inner.to_string(),
            LiteralValue::Int(inner) => inner.to_string(),
            LiteralValue::UInt(inner) => inner.to_string(),
            LiteralValue::String(inner) => format!("'{inner}'"),
        }
    }

    fn wrap_in_comment(
        &self,
        comment: Option<&String>,
        default: Option<&LiteralValue>,
        value: String,
    ) -> String {
        let indent = self.indent();
        let mut lines = vec![];

        if let Some(comment) = comment {
            lines.extend(
                comment
                    .trim()
                    .split('\n')
                    .map(|c| c.trim().to_owned())
                    .collect::<Vec<_>>(),
            );
        }

        if let Some(default) = default {
            lines.push(format!("@default {}", self.lit_to_string(default)));
        }

        if lines.is_empty() {
            return value;
        }

        if lines.len() == 1 {
            return format!("{}/** {} */\n{}", indent, lines[0], value);
        }

        let mut out = vec![format!("{}/**", indent)];

        for line in lines {
            out.push(format!("{} * {}", indent, line.trim()));
        }

        out.push(format!("{} */", indent));

        format!("{}\n{}", out.join("\n"), value)
    }
}

impl SchemaRenderer<String> for TypeScriptRenderer {
    fn is_reference(&self, name: &str) -> bool {
        if self.options.disable_references {
            return false;
        }

        if self.references.contains(name) {
            return true;
        }

        self.is_external(name)
    }

    fn render_array(&mut self, array: &ArrayType) -> RenderResult {
        let out = self.render_schema(&array.items_type)?;

        Ok(if out.contains('|') {
            format!("({})[]", out)
        } else {
            format!("{}[]", out)
        })
    }

    fn render_boolean(&mut self, _boolean: &BooleanType) -> RenderResult {
        Ok("boolean".into())
    }

    fn render_enum(&mut self, enu: &EnumType) -> RenderResult {
        self.render_enum_or_union(enu)
    }

    fn render_float(&mut self, float: &FloatType) -> RenderResult {
        if let Some(values) = &float.enum_values {
            return Ok(values
                .iter()
                .map(|v| v.to_string())
                .collect::<Vec<_>>()
                .join(" | "));
        }

        Ok("number".into())
    }

    fn render_integer(&mut self, integer: &IntegerType) -> RenderResult {
        if let Some(values) = &integer.enum_values {
            return Ok(values
                .iter()
                .map(|v| v.to_string())
                .collect::<Vec<_>>()
                .join(" | "));
        }

        Ok("number".into())
    }

    fn render_literal(&mut self, literal: &LiteralType) -> RenderResult {
        if let Some(value) = &literal.value {
            return Ok(self.lit_to_string(value));
        }

        self.render_unknown()
    }

    fn render_null(&mut self) -> RenderResult {
        Ok("null".into())
    }

    fn render_object(&mut self, object: &ObjectType) -> RenderResult {
        Ok(format!(
            "Record<{}, {}>",
            self.render_schema(&object.key_type)?,
            self.render_schema(&object.value_type)?
        ))
    }

    fn render_reference(&mut self, reference: &str) -> RenderResult {
        Ok(reference.into())
    }

    fn render_string(&mut self, string: &StringType) -> RenderResult {
        if let Some(values) = &string.enum_values {
            return Ok(values
                .iter()
                .map(|v| format!("'{}'", v))
                .collect::<Vec<_>>()
                .join(" | "));
        }

        Ok("string".into())
    }

    fn render_struct(&mut self, structure: &StructType) -> RenderResult {
        self.depth += 1;

        let mut out = vec![];
        let indent = self.indent();

        for field in &structure.fields {
            if field.hidden {
                continue;
            }

            let mut row = format!("{}{}", indent, field.name.as_ref().unwrap());

            if field.optional {
                row.push_str("?: ");
            } else {
                row.push_str(": ");
            }

            row.push_str(&self.render_schema(&field.type_of)?);

            if matches!(self.options.object_format, ObjectFormat::Interface) {
                row.push(';');
            } else {
                row.push(',');
            }

            row =
                self.wrap_in_comment(field.description.as_ref(), field.type_of.get_default(), row);

            out.push(row);
        }

        self.depth -= 1;

        Ok(format!("{{\n{}\n{}}}", out.join("\n"), self.indent()))
    }

    fn render_tuple(&mut self, tuple: &TupleType) -> RenderResult {
        let mut items = vec![];

        for item in &tuple.items_types {
            items.push(self.render_schema(item)?);
        }

        Ok(format!("[{}]", items.join(", ")))
    }

    fn render_union(&mut self, uni: &UnionType) -> RenderResult {
        let mut items = vec![];

        for item in &uni.variants_types {
            items.push(self.render_schema(item)?);
        }

        Ok(items.join(" | "))
    }

    fn render_unknown(&mut self) -> RenderResult {
        Ok("unknown".into())
    }

    fn render(
        &mut self,
        schemas: &IndexMap<String, SchemaType>,
        references: &HashSet<String>,
    ) -> RenderResult {
        self.references.extend(references.to_owned());

        let mut outputs = vec![
            "// Automatically generated by schematic. DO NOT MODIFY!".to_string(),
            "/* eslint-disable */".to_string(),
        ];

        let mut imports = vec![];

        for (import, types) in &self.options.external_types {
            let mut imported_types = types.iter().cloned().collect::<Vec<_>>();
            imported_types.sort();

            imports.push(format!(
                "import type {{ {} }} from '{}';",
                imported_types.join(", "),
                import
            ));
        }

        if !imports.is_empty() {
            outputs.push(imports.join("\n"));
        }

        for (name, schema) in schemas {
            if self.is_excluded(name) {
                continue;
            }

            outputs.push(match schema {
                SchemaType::Enum(inner) => self.export_enum_type(name, inner)?,
                SchemaType::Struct(inner) => self.export_object_type(name, inner)?,
                _ => {
                    let out = self.render_schema_without_reference(schema)?;
                    self.export_type_alias(name, out)?
                }
            });
        }

        Ok(outputs.join("\n\n"))
    }
}
