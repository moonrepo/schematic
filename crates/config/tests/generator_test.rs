#![allow(dead_code, deprecated)]

use schematic::schema::SchemaGenerator;
use schematic::*;
use starbase_sandbox::{assert_snapshot, create_empty_sandbox};
use std::collections::{HashMap, HashSet};
use std::fs;

derive_enum!(
    #[derive(ConfigEnum, Default)]
    pub enum BasicEnum {
        #[default]
        Foo,
        Bar,
        Baz,
    }
);

#[derive(Clone, Config)]
pub struct AnotherConfig {
    opt: Option<String>,
    enums: Option<BasicEnum>,
}

#[derive(Clone, Config)]
struct GenConfig {
    boolean: bool,
    string: String,
    number: usize,
    vector: Vec<String>,
    map: HashMap<String, u64>,
    enums: BasicEnum,
    #[setting(nested)]
    nested: AnotherConfig,
}

fn create_generator() -> SchemaGenerator {
    let mut generator = SchemaGenerator::default();
    generator.add::<GenConfig>();
    generator
}

#[cfg(feature = "typescript")]
mod typescript {
    use super::*;
    use schematic::renderers::typescript::*;

    fn generate(options: TypeScriptOptions) -> String {
        let sandbox = create_empty_sandbox();
        let file = sandbox.path().join("types.ts");

        create_generator()
            .generate(&file, TypeScriptRenderer::new(options))
            .unwrap();

        fs::read_to_string(file).unwrap()
    }

    #[test]
    fn defaults() {
        assert_snapshot!(generate(TypeScriptOptions::default()));
    }

    #[test]
    fn enums() {
        assert_snapshot!(generate(TypeScriptOptions {
            enum_format: EnumFormat::Enum,
            ..TypeScriptOptions::default()
        }));
    }

    #[test]
    fn value_enums() {
        assert_snapshot!(generate(TypeScriptOptions {
            enum_format: EnumFormat::ValuedEnum,
            ..TypeScriptOptions::default()
        }));
    }

    #[test]
    fn const_enums() {
        assert_snapshot!(generate(TypeScriptOptions {
            const_enum: true,
            enum_format: EnumFormat::Enum,
            ..TypeScriptOptions::default()
        }));
    }

    #[test]
    fn object_aliases() {
        assert_snapshot!(generate(TypeScriptOptions {
            object_format: ObjectFormat::Type,
            ..TypeScriptOptions::default()
        }));
    }

    #[test]
    fn exclude_refs() {
        assert_snapshot!(generate(TypeScriptOptions {
            exclude_references: HashSet::from_iter(["BasicEnum".into(), "AnotherType".into()]),
            ..TypeScriptOptions::default()
        }));
    }

    #[test]
    fn external_types() {
        assert_snapshot!(generate(TypeScriptOptions {
            external_types: HashMap::from_iter([(
                "./externals".into(),
                HashSet::from_iter(["BasicEnum".into(), "AnotherType".into()])
            )]),
            ..TypeScriptOptions::default()
        }));
    }

    #[test]
    fn no_refs() {
        assert_snapshot!(generate(TypeScriptOptions {
            disable_references: true,
            indent_char: "  ".into(),
            ..TypeScriptOptions::default()
        }));
    }
}
