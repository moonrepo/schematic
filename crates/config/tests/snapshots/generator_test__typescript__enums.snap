---
source: crates/config/tests/generator_test.rs
expression: "generate(TypeScriptOptions {\n        enum_format: EnumFormat::Enum,\n        ..TypeScriptOptions::default()\n    })"
---
// Automatically generated by schematic. DO NOT MODIFY!

/* eslint-disable */

export enum BasicEnum {
	Foo,
	Bar,
	Baz,
}

export interface AnotherConfig {
	enums: BasicEnum | null;
	opt: string | null;
}

export interface GenConfig {
	boolean: boolean;
	date: string;
	datetime: string;
	enums: BasicEnum;
	float32: number;
	float64: number;
	id: string;
	locator: string | null;
	map: Record<string, number>;
	nested: AnotherConfig;
	number: number;
	path: string;
	relPath: string;
	spec: string;
	specUnresolved: string;
	string: string;
	time: string;
	url: string | null;
	vector: string[];
	version: string | null;
	versionReq: string;
}

